pkg/cmd/server/cacheconfig.go

```diff
@@ -2,64 +2,79 @@ package server
-func (cc *CacheConfig) Complete() (*cache.Config, error) {
-	if cc.Disabled {
-		return &cache.Config{
-			MaxCost:     1,
-			NumCounters: cc.NumCounters,
-			Metrics:     cc.Metrics,
-			BufferItems: defaultBufferItems,
-			Disabled:    true,
-		}, nil
+func (cc *CacheConfig) Complete() (cache.Cache, error) {
+	if !cc.Enabled || cc.MaxCost == "" || cc.NumCounters == 0 {
+		return cache.NoopCache(), nil
 	}
 
-	if cc.MaxCost == "" || cc.NumCounters == 0 {
-		return nil, nil
-	}
+	var (
+		maxCost uint64
+		err     error
+	)
 
-	maxCost, err := humanize.ParseBytes(cc.MaxCost)
+	if strings.HasSuffix(cc.MaxCost, "%") {
+		maxCost, err = parsePercent(cc.MaxCost)
+	} else {
+		maxCost, err = humanize.ParseBytes(cc.MaxCost)
+	}
 	if err != nil {
-		return nil, fmt.Errorf("error parsing cache max cost `%s`: %w", cc.MaxCost, err)
+		return nil, fmt.Errorf("error parsing cache max memory: `%s`: %w", cc.MaxCost, err)
 	}
 
-	return &cache.Config{
+	return cache.NewCache(&cache.Config{
 		MaxCost:     int64(maxCost),
 		NumCounters: cc.NumCounters,
 		Metrics:     cc.Metrics,
-		BufferItems: defaultBufferItems,
-		Disabled:    false,
-	}, nil
+	})
+}
+
+func parsePercent(str string) (uint64, error) {
+	percent := strings.TrimSuffix(str, "%")
+	parsedPercent, err := strconv.ParseUint(percent, 10, 64)
+	if err != nil {
+		return 0, fmt.Errorf("failed to parse percentage: %w", err)
+	}
+
+	if parsedPercent > 100 {
+		return 0, fmt.Errorf("percentage greater than 100")
+	}
+
+	return freeMemory / 100 * parsedPercent, nil
 }
 
-// RegisterCacheConfigFlags registers flags for a ristretto-based cache.
-func RegisterCacheConfigFlags(flags *pflag.FlagSet, config *CacheConfig, flagPrefix string) {
+// RegisterCacheFlags registers flags used to configure SpiceDB's various
+// caches.
+func RegisterCacheFlags(flags *pflag.FlagSet, flagPrefix string, config, defaults *CacheConfig) {
 	flagPrefix = stringz.DefaultEmpty(flagPrefix, "cache")
-	flags.StringVar(&config.MaxCost, flagPrefix+"-max-cost", defaultMaxCost, "the maximum cost to be stored in the cache, in bytes")
-	flags.Int64Var(&config.NumCounters, flagPrefix+"-num-counters", defaultNumCounters, "the number of keys to track")
-	flags.BoolVar(&config.Metrics, flagPrefix+"-metrics", false, "whether metrics should be maintained for the cache. WARNING: Incurs a performance penality.")
-	flags.BoolVar(&config.Disabled, flagPrefix+"-disabled", false, "if true, fully disables the cache. WARNING: Incurs a performance penality.")
+	flags.StringVar(&config.MaxCost, flagPrefix+"-max-cost", defaults.MaxCost, "upper bound cache size in bytes or percent of available memory")
+	flags.Int64Var(&config.NumCounters, flagPrefix+"-num-counters", defaults.NumCounters, "number of TinyLFU samples to track")
+	flags.BoolVar(&config.Metrics, flagPrefix+"-metrics", false, "enable cache metrics")
+	flags.BoolVar(&config.Enabled, flagPrefix+"-enabled", true, "enable caching")
```

Comment:
There two parameters are ignoring the provided default:

```suggestion
	flags.BoolVar(&config.Metrics, flagPrefix+"-metrics", defaults.Metrics, "enable cache metrics")
	flags.BoolVar(&config.Enabled, flagPrefix+"-enabled", defaults.Enabled, "enable caching")
```